jwt/jwt.go

```diff
@@ -0,0 +1,202 @@
+	sess, err := session.NewSession()
+	if err != nil {
+		return "", err
+	}
+
+	svc := ssm.New(sess)
+	input := &ssm.GetParameterInput{
+		Name:           aws.String(parameterName),
+		WithDecryption: aws.Bool(true),
+	}
+
+	result, err := svc.GetParameter(input)
+	if err != nil {
+		return "", err
+	}
+
+	return *result.Parameter.Value, nil
+}
+
+func (j *JWTUtils) ValidateToken(tokenString string) (jwt.MapClaims, error) {
+	if j == nil || j.publicKey == nil {
+		return nil, errors.New("internal server error")
+	}
+
+	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("invalid token")
+		}
+		return j.publicKey, nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("invalid token")
+	}
+
+	if !token.Valid {
+		return nil, errors.New("invalid token")
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, errors.New("invalid token")
+	}
+
+	return claims, nil
+}
+
+func (j *JWTUtils) ExtractClaim(claims jwt.MapClaims, claimKey string) (string, error) {
+	if claims == nil {
+		return "", errors.New("internal server error")
+	}
+
+	value, ok := claims[claimKey].(string)
+	if !ok || value == "" {
+		return "", fmt.Errorf("unauthorized")
+	}
+	return value, nil
+}
+
+func JWTMiddleware() func(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, string, error) {
+	return func(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, string, error) {
+		jwtUtils, err := GetInstance()
+		if err != nil {
+			log.Printf("Failed to get JWTUtils instance: %v", err)
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusInternalServerError,
+				Body:       "Internal server error",
+			}, "", nil
+		}
+
+		cookie := req.Headers["Cookie"]
+		if cookie == "" {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       "Unauthenticated",
+			}, "", nil
+		}
+
+		cookieName := os.Getenv("SESSION_COOKIE_NAME")
+		if cookieName == "" {
+			cookieName = utils.DEVELOPMENT_COOKIE_NAME
+
+		}
+		var jwtToken string
+		cookies := strings.Split(cookie, ";")
+		for _, c := range cookies {
+			c = strings.TrimSpace(c)
+			if strings.HasPrefix(c, cookieName+"=") {
+				jwtToken = strings.TrimPrefix(c, cookieName+"=")
+				break
+			}
+		}
+
+		if jwtToken == "" {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       "Unauthenticated",
+			}, "", nil
+		}
+
+		claims, err := jwtUtils.ValidateToken(jwtToken)
+		if err != nil {
+			log.Printf("Token validation failed: %v", err)
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       "Invalid token",
+			}, "", nil
+		}
+
+		userId, err := jwtUtils.ExtractClaim(claims, "userId")
+		if err != nil {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       "Unauthorized",
```

Comment:
- should this be `invalid token` instead of `unauthorized`?

Reply:
I removed this because we were previously using the v2 cookie, which included a userID. The v1 cookie does not contain the userID, so I removed the claim as it is no longer relevant.

Reply:
Umm I have changed the error message to `User ID claim could not be extracted, Unauthorized ` , We already have a invalid token message when there is a error while validating the token. and here we are trying to extract the UserId from token. so Since this error specifically occurs when extracting the User ID from the token, the `Invalid token` message didn't seem clear error message.